gtfs-realtime-validator-lib/src/main/java/edu/usf/cutr/gtfsrtvalidator/lib/validation/rules/FrequencyTypeZeroValidator.java

```diff
@@ -75,9 +83,25 @@
                         // W005 - Missing vehicle_id in trip_update for frequency-based exact_times = 0
                         RuleUtils.addOccurrence(W005, "trip_id " + tripUpdate.getTrip().getTripId(), errorListW005, _log);
                     }
+
+                    if (tripUpdate.hasVehicle() && tripUpdate.getVehicle().hasId() && previousTripUpdates.get(tripUpdate.getVehicle().getId()) != null) {
```

Comment:
Thinking more about this in context of TripUpdates, I think that we need to use a composite key for the HashMap - `vehicle_id` + `trip_id`.

In a TripUpdates feed, you can have multiple TripUpdate records that have the same vehicle_id.  For example, if a block is composed of two trips:
* trip_A
* trip_B

You can have a TripUpdate for trip_A and a TripUpdate for trip_B in the feed at the same time for the same vehicle if the same vehicle is serving both trips in the block.

All exact_times=0 trips that I've seen in real GTFS datasets have been loops, so they don't have the above issue (they just repeat the same trip_id over and over).  But, I believe it's possible to have two trips defined in trips.txt and frequencies.txt that would have two different frequencies for exact_times=0 trips that would be served by the same vehicle back-to-back.

@scrudden Let me know what you think of the above.

Reply:
You can also have two trip updates in a single feed with the same trip id/vehicle id and different start times and be correct.

# Previous updates vehicle 1001
Trip_A_08:00
Trip_A_09:00

# Possible current updates vehicle 1001
## 1 Updates for same trips as last update
Trip_A_08:00
Trip_A_09:00

## 2 Trip_A_08:00 completed so has updated for next two trips
Trip_A_09:00 
Trip_A_10:00

So we need logic for these two scenarios. 1 is simple as if all match then OK. 2 Trip_A_10:00 is OK but how do we know?

Do we check that the start time (10:00) is after the latest matching start_time that is in both feeds for this vehicle?

Reply:
If there are 25 stops in a trip, and we assign a trip_instance value in this example just to keep track of them in our discussion, the TripUpdate data would probably look something like:

# Previous updates vehicle 1001
Trip_A_08:00 - last stop_sequence = 20 (trip_instance=0)
Trip_A_09:00 - last stop_sequence  = 10 (trip_instance=1)

# Possible current updates vehicle 1001
## 1 Updates for same trips as last update
Trip_A_08:00 - last stop_sequence = 25 (trip_instance=0)
Trip_A_09:00 - last stop_sequence = 15 (trip_instance=1)

## 2 Trip_A_08:00 completed so has updated for next two trips
Trip_A_09:00  - last stop_sequence = 20 (trip_instance=1)
Trip_A_10:00  - last stop_sequence = 5 (trip_instance=2)

So for scenario 2, you'd be comparing:

*  Trip_A_09:00  - last stop_sequence = 20 (trip_instance=1) against
    * Trip_A_08:00 - last stop_sequence = 20 (trip_instance=0) // This would incorrectly log an error
    * Trip_A_09:00 - last stop_sequence  = 10 (trip_instance=1) // This correctly does not log an error
* Trip_A_10:00  - last stop_sequence = 5 (trip_instance=2) against
    * Trip_A_08:00 - last stop_sequence = 20 (trip_instance=0) // This correctly does not log an error
    * Trip_A_09:00 - last stop_sequence  = 10 (trip_instance=1) // This correctly does not log an error

I think I got that right?

One issue with the above, though is that currently my suggestion of using `vehicle_id + trip_id` -> TripUpdate as a composite key would create collisions in a HashMap when executing scenario 2.  So we'd have to create a HashMap of `vehicle_id + trip_id` -> List<TripUpdate> or something else to even be able to track this in code.

It would definitely be helpful to define these scenarios in the unit tests to see which ones we're covering and which we aren't.  I'm fine with covering the cases we can, and then in the rule description acknowledging which cases we may have false positive or false negatives on.

Reply:
>The last stop sequence will be the last stop of the trip if there is an update for a subsequent trip.

That's not guaranteed, but I agree that will be typical behavior.

>I have already locally added a sorted list of trip updates to the HashMap for the current and previous messages

For the strategy I was thinking of, I don't think we need a HashMap for the current message - we should just be able loop through the current message once.  I'm not sure I understand why you're sorting by start_time, either - if possible I'd prefer to avoid sorts as we pay another runtime penalty of at least log(n).

>I think adding tests to match the above scenarios will be a start to this.

I agree, that's the only way I'll be able to convince myself it works :).  I had the same issue with some other tests that had a number of different permutations of field values.

>I'll check in what I have changed but it is very much a WIP.

Ok, sounds good.  Let me know when you want me to go through it again.